### PR TITLE
Abstract the disk/lvol file operations in restore functions

### PR DESCRIPTION
**Pull request by derekbit**
https://github.com/longhorn/longhorn/issues/6576

### CODE REVIEW

File: deltablock.go [hunk 1]
```diff
@@ -736,8 +729,8 @@ func RestoreDeltaBlockBackup(config *DeltaRestoreConfig) error {
 		// closed.
 		// https://github.com/longhorn/longhorn/issues/2503
 		// We want to truncate regular files, but not device
-		if stat.Mode()&os.ModeType == 0 {
-			log.Debugf("Truncate %v to size %v", volDevName, vol.Size)
+		if stat.Mode().IsRegular() && stat.Mode()&os.ModeType == 0 {
```
Comment (innobead) [inline deltablock.go:735]: We can just remove the duplicated `stat.Mode()&os.ModeType == 0` from this if statement. 

```go
func (m FileMode) IsRegular() bool {
	return m&ModeType == 0
}
```

**Review by innobead — CHANGES_REQUESTED** [review]
LGTM, but some comments need to resolve. 

The dependant component repos need to be updated and backported.

File: deltablock.go (hunk 1 above)
Comment (derekbit) [reply deltablock.go]: Updated. Thank you.

File: types.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+	BLOCKS_DIRECTORY      = "blocks"
+	BLOCK_SEPARATE_LAYER1 = 2
+	BLOCK_SEPARATE_LAYER2 = 4
+	BLK_SUFFIX            = ".blk"
+
+	PROGRESS_PERCENTAGE_BACKUP_SNAPSHOT = 95
+	PROGRESS_PERCENTAGE_BACKUP_TOTAL    = 100
+
+	BackendStoreDriverV1 = "v1"
+	BackendStoreDriverV2 = "v2"
```
[the remark this answers is not in the document]
Comment (derekbit) [reply types.go]: Addressed. Thank you.

**Comment by derekbit** [comment]
cc @innobead

File: deltablock.go [hunk 3]
```diff
@@ -362,7 +373,7 @@ func backupBlock(bsDriver BackupStoreDriver, config *DeltaBackupConfig,
 		return nil
 	}
 
-	log.Debugf("Creating new block file at %v", blkFile)
+	log.Tracef("Creating new block file at %v", blkFile)
```
Comment (innobead) [inline deltablock.go:376]: If we can apply the log level setting of longhorn-manager to indirectly apply to all components, then debug level should be enough.

File: deltablock.go [hunk 4]
```diff
@@ -1069,7 +1050,7 @@ func performIncrementalRestore(bsDriver BackupStoreDriver, config *DeltaRestoreC
 
 	errorChans := []<-chan error{errChan}
 	for i := 0; i < int(concurrentLimit); i++ {
-		errorChans = append(errorChans, restoreBlocks(ctx, bsDriver, config, srcVolumeName, blockChan, progress))
+		errorChans = append(errorChans, restoreBlocks(ctx, bsDriver, config.DeltaOps, config.Filename, srcVolumeName, blockChan, progress))
```
Comment (innobead) [inline deltablock.go:1039]: So incremental restore will be supported later? or we need to plan for 1.7 instead?

File: deltablock.go [hunk 5]
```diff
@@ -1018,32 +976,41 @@ func restoreBlock(bsDriver BackupStoreDriver, config *DeltaRestoreConfig,
...
+		}()
 
 		for {
 			select {
 			case <-ctx.Done():
+				logrus.Infof("Closing goroutine for restoring blocks for volume %v", volumeName)
+				return
+			case <-deltaOps.GetStopChan():
+				logrus.Infof("Closing goroutine for restoring blocks for %v since received stop signal", volumeName)
+				err = fmt.Errorf("restoration is cancelled since received stop signal")
```
Comment (innobead) [inline deltablock.go:1005]: nit: Suggest adding volume name in the error message to ensure the complete error context, even though the caller could enrich the context (volume name) already at the upper layer.

**Review by innobead — APPROVED** [review]
LGTM. Just a few comments, but it's good to merge.

File: deltablock.go (hunk 3 above)
Comment (innobead) [reply deltablock.go:376]: ref: https://github.com/longhorn/longhorn/issues/6702

**Pull request merged**